### DISCUSSION TITLE
Is it possible to get social cards to work with mkdocs-static-i18n?

### DISCUSSION THREAD

**Discussion by alexjlockwood** [squidfunk/mkdocs-material#5592] [Q&A]
Referring to this translation plugin here: https://github.com/ultrabug/mkdocs-static-i18n

Basically, the plugin allows you to place markdown files side by side (i.e. `index.md` for English and `index.ko.md` for Korean) and it will create a separate translated version of the page at a different link. However, when I try to set up social cards, it seems like the Korean translated social card is simply overwriting the English translated social card.

I'm wondering if there is any workaround here that would allow me to create both an English social card and a Korean social card? Or any examples that you may know of to get it working?

Also, a similar question I have is whether it is possible to have the English translation use one Font, and Korean translation use a different font? Because my Korean social cards are not rendering properly with the current default font because it doesn't support Korean characters...

As a last resort, if none of this is possible, I'm wondering if there is a way to simply turn off social card rendering for the translations (i.e. just have them default to the English version, even though it is not ideal it is better than nothing).

**Comment by squidfunk** [comment]
> However, when I try to set up social cards, it seems like the Korean translated social card is simply overwriting the English translated social card.

From reading the source code, the mkdocs-static-i18n plugin seems to do a lot of legwork to [bring other third-party plugins under its control](https://github.com/ultrabug/mkdocs-static-i18n/blob/67a324bd019d9a0f3b31afb652e06b8e0af64095/mkdocs_static_i18n/plugin.py#L68-L70). I'm not sure whether the social  plugin requires the same thing. Guessing, I would say no, because the social plugin will just generate a social card for every page, and if you have two separate pages, those should be two separate social cards. Thus, it might be a better idea to ask the author of the mkdocs-static-i18n plugin, because from our side, it should work without problems. This should apply to the brand new social plugin that is distributed with Insiders, as well as to the version we currently have in the community edition. Feel free to come back if you gather any insights! ☺️

> Also, a similar question I have is whether it is possible to have the English translation use one Font, and Korean translation use a different font? Because my Korean social cards are not rendering properly with the current default font because it doesn't support Korean characters...

Definitely possible with the new social plugin! You can set the font in a custom social card template by examining the `page` object (possibly check if `page.src_uri` ends in `ko.md`). See the [`default` template](https://github.com/squidfunk/mkdocs-material-insiders/blob/2a1023da4c55f4115ffb77507ed9ea6cb90266e0/src/plugins/social/layouts/default.yml#L104-L112) to learn how we also set the font family conditionally.

Alternatively, you can subscribe to #5567, in which it was already requested to change the social card template per page, albeit [only in this comment](https://github.com/squidfunk/mkdocs-material/issues/5567#issuecomment-1569313961). If you wish, you can open a PR that requests that functionality specifically. Otherwise, we'll add it anyway in the future.

**Reply by squidfunk** [reply]
If you would like to set the font family, see [this YouTube video](https://www.youtube.com/watch?v=4OjnOc6ftJ8) for an excellent tutorial on creating custom social cards.

**Reply by alexjlockwood** [reply]
Thanks for the reply! I've filed an issue against the internationalization plugin here: https://github.com/ultrabug/mkdocs-static-i18n/issues/228

Hopefully they respond with some advice on what to do...

**Reply by squidfunk** [reply]
Regarding your font question, [I found this custom template in the wild](https://github.com/privacyguides/privacyguides.org/blob/ef7bd365ccc6a173958d694741da1fbb95d2d5c7/config/layouts/custom.yml#L23-L28) ☺️

``` yaml
  - &title_font_family >-
    {%- if config.theme.language == "he" -%}
      Suez One
    {%- else -%}
      Bagnard
    {%- endif -%}
```

Note that this is only an excerpt.

**Comment by kamilkrzyskow** [comment]
You could check out my message here: https://github.com/squidfunk/mkdocs-material/discussions/5359#discussioncomment-5812615 
I made changes to support the social plugin too.

The big refactor was supposed to fix most of such issues, but the PR got stale, and I believe the author didn't want to accept other PRs before completing his work on it.

**Reply by squidfunk** [reply]
This will definitely not work with the new social plugin that is in Insiders, [as this attribute does not exist anymore](https://github.com/kamilkrzyskow/i18n/commit/baabb44d77918aae5bb4870878716c4dd87a546b#diff-326a6088432fa7937bc18d7583396cab622b3cc378c3840713652bcb3b5e0b55R652). I'm not sure whether this approach scales well, because it seems the i18n plugin needs exceptions for many plugins.

**Reply by kamilkrzyskow** [reply]
Weird, you mean that somehow `src_uri` and `src_path` of the `page` object aren't available? I thought they're built-in in MkDocs. GitHub on Android doesn't show me any highlighted line. Then again I don't know what you mean it would not work in Insiders, probably you meant that the attribute isn't used anymore or something else with the promises.

This was a bandaid fix to get it working. A change in the social plugin's API could easily break it. The refactor aimed to minimize such occurrences via rerunning the whole event loop again for each language, though I recall that it would still not work well, because the path for the reused pages would overwrite the social images, same as the OP post.

That's the sad truth about the i18n plugin. MkDocs is built around the idea that a single config file matches one language, this influences other plugins doing the same, and the i18n tries to make it easier for the user by breaking the rules of everything underneath.

Anyways, I hope that OP will figure it out, a simple hook should be enough to override the default behaviour. At least I think so without seeing the code.

**Reply by alexjlockwood** [reply]
I'm not so sure on how to proceed. Seems like I just might not be able to use social card plugin for now, which is a shame because it looked really awesome! Unfortunately I don't know much about mkdocs internals and how plugins work so I doubt I would be able to figure out some sort of workaround myself. And sounds like the fact that I am using mkdocs-material insiders will make that more difficult to figure out.

If any of you have any alternate suggestions in terms of i18n plugins, I'm open to hearing them. Although mkdocs-static-i18n really seemed like the best option when I was researching alternatives.

**Reply by squidfunk** [reply]
I recommend reading https://github.com/squidfunk/mkdocs-material/discussions/2346, which explains how to achieve a multi-language installation with separate MkDocs projects ☺️ The ground up rewrite of the social plugin in Insiders was really necessary to make it more powerful, and yes, the i18n plugins seems to use internals of the community version, so it will not work with Insiders.

> Weird, you mean that somehow src_uri and src_path of the page object aren't available? I thought they're built-in in MkDocs. GitHub on Android doesn't show me any highlighted line. Then again I don't know what you mean it would not work in Insiders, probably you meant that the attribute isn't used anymore or something else with the promises.

My comment was solely related to the promise attribute. The new version in Insiders doesn't have that attribute.
